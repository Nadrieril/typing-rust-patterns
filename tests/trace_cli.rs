use assert_cmd::Command;
use std::collections::hash_map::DefaultHasher;
use std::hash::{Hash, Hasher};

#[test]
fn trace_cli() -> anyhow::Result<()> {
    let test_cases = &[
        "quit",
        "help",
        "options",
        "set",
        "set mistyped",
        "set mistyped val",
        "set ref_binding_on_inherited true",
        "set allow_ref_pat_on_ref_mut false",
        "set downgrade_shared_inside_shared true",
        "set ergo2024\noptions",
        "&x: &mut T",
        "set allow_ref_pat_on_ref_mut false\n&x: &mut T",
        "rules",
    ];
    for &test_case in test_cases {
        let success = Command::cargo_bin("typing-rust-patterns")?
            .write_stdin(test_case)
            .assert()
            .success();
        let out = String::from_utf8(success.get_output().stdout.clone())?;

        // Identify each snapshot file by the hash of the test case.
        let req_hash = {
            let mut hasher = DefaultHasher::new();
            test_case.hash(&mut hasher);
            hasher.finish().to_string()
        };
        insta::with_settings!({
            snapshot_suffix => req_hash,
            info => &test_case,
            omit_expression => true,
            prepend_module_to_snapshot => true,
        }, {
            insta::assert_snapshot!(out);
        });
    }
    Ok(())
}
