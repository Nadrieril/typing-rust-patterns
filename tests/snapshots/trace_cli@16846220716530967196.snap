---
source: tests/trace_cli.rs
info: "set rules_display_style Sequent\nrules\n&[[x]]: &[&mut [x]]"
---
The current options can be fully described as the following set of rules.
The typing predicate looks like `b, m ⊢ p: T`, where
- `b` is `place` or `value` and indicates the place context aka binding mode;
- `m` is `rw` or `ro` and indicates whether we have mutable or read-only access to the original scrutinee;
- `p` is a pattern;
- `T` is a type.

place, m ⊢ p0: T0,  place, m ⊢ p1: T1
----------------------------------------- "Constructor"
place, m ⊢ [p0, p1]: [T0, T1]

value, ro ⊢ p0: &T0,  value, ro ⊢ p1: &T1
--------------------------------------------- "ConstructorRef"
b, m ⊢ [p0, p1]: &[T0, T1]

value, m ⊢ p0: &mut T0,  value, m ⊢ p1: &mut T1
--------------------------------------------------- "ConstructorRef"
b, m ⊢ [p0, p1]: &mut [T0, T1]

value, ro ⊢ [p0, p1]: &T
-------------------------- "ConstructorMultiRef"
b, m ⊢ [p0, p1]: &&T

value, ro ⊢ [p0, p1]: &T
-------------------------- "ConstructorMultiRef"
b, m ⊢ [p0, p1]: &&mut T

value, ro ⊢ [p0, p1]: &T
-------------------------- "ConstructorMultiRef"
b, m ⊢ [p0, p1]: &mut &T

value, m ⊢ [p0, p1]: &mut T
------------------------------ "ConstructorMultiRef"
b, m ⊢ [p0, p1]: &mut &mut T

place, ro ⊢ p: T
------------------ "Deref"
b, m ⊢ &p: &T

place, m ⊢ p: T
----------------------- "Deref"
b, m ⊢ &mut p: &mut T

place, ro ⊢ p: T
------------------- "DerefMutWithShared"
b, m ⊢ &p: &mut T

value, m ⊢ x: &T
--------------------- "BindingBorrow"
place, m ⊢ ref x: T

value, m ⊢ x: &mut T
------------------------- "BindingBorrow"
place, m ⊢ ref mut x: T


------------- "Binding"
b, m ⊢ x: T


----------------- "Binding"
b, m ⊢ mut x: T

place, rw ⊢ &[[x]]: &[&mut [x]]
// Applying rule `Deref`
place, ro ⊢ [[x]]: [&mut [x]]
// Applying rule `Constructor`
place, ro ⊢ [x]: &mut [x]
// Applying rule `ConstructorRef`
value, ro ⊢ x: &mut x
// Applying rule `Binding`
let x: &mut x = &mut (*(*s).0).0

// Final bindings (simplified):
let x: &mut x = &mut (*(*s).0).0; // Borrow-check error: MutBorrowBehindSharedBorrow
